src/renderers/shared/fiber/ReactChildFiber.js

```diff
@@ -16,224 +16,768 @@ import type { ReactCoroutine, ReactYield } from 'ReactCoroutine';
+    // just store this on the fiber itself since that lets us use the "implicit"
+    // index resolution mechanism without adding null values to the linked list.
+    let existingChild = currentFirstChild;
+    while (existingChild) {
+      if (existingChild.key !== null) {
+        existingChildren.set(existingChild.key, existingChild);
+      }
+      // Add everything to the delete queue
+      // Actually... It is not possible to delete things from the queue since
+      // we don't have access to the previous link. Does that mean we need a
+      // second pass to add them? We should be able to keep track of the
+      // previous deletion as we're iterating through the list the next time.
+      // That way we know which item to patch when we delete a deletion.
+      existingChild = existingChild.sibling;
+    }
+    return existingChildren;
+  }
+
+  function useFiber(fiber : Fiber, priority : PriorityLevel) {
+    // We currently set sibling to null and index to 0 here because it is easy
+    // to forget to do before returning it. E.g. for the single child case.
+    if (shouldClone) {
+      const clone = cloneFiber(fiber, priority);
+      clone.index = 0;
+      clone.sibling = null;
+      return clone;
+    } else {
+      if (fiber.pendingWorkPriority === NoWork ||
+          fiber.pendingWorkPriority > priority) {
+        fiber.pendingWorkPriority = priority;
       }
+      fiber.index = 0;
+      fiber.sibling = null;
+      return fiber;
+    }
+  }
 
-      case REACT_COROUTINE_TYPE: {
-        const coroutine = (newChildren : ReactCoroutine);
-        const child = createFiberFromCoroutine(coroutine, priority);
-        previousSibling.sibling = child;
-        child.return = returnFiber;
-        return child;
+  function placeChild(newFiber : Fiber, lastPlacedIndex : number, newIndex : number) {
+    newFiber.index = newIndex;
+    if (!shouldTrackSideEffects) {
+      // Noop.
+      return lastPlacedIndex;
+    }
+    const current = newFiber.alternate;
+    if (current) {
+      const oldIndex = current.index;
+      if (oldIndex < lastPlacedIndex) {
+        // This is a move.
+        // TODO: Schedule a move side-effect for this child.
+        return lastPlacedIndex;
+      } else {
+        // This item can stay in place.
+        return oldIndex;
       }
+    } else {
+      // This is an insertion.
+      // TODO: Schedule an insertion side-effect for this child.
+      return lastPlacedIndex;
+    }
+  }
+
+  function updateTextNode(
+    returnFiber : Fiber,
+    current : ?Fiber,
+    textContent : string,
+    priority : PriorityLevel
+  ) {
+    if (current == null || current.tag !== HostText) {
+      // Insert
+      const created = createFiberFromText(textContent, priority);
+      created.return = returnFiber;
+      return created;
+    } else {
+      // Update
+      const existing = useFiber(current, priority);
+      existing.pendingProps = textContent;
+      existing.return = returnFiber;
```

Comment:
It looks like whenever `useFiber` is used, you also set the return fiber. Any reason this is kept separate?